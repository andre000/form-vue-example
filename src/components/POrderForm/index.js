export { default as POrderForm } from './POrderForm.vue'
