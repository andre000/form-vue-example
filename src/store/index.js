import Vue from 'vue'
import Vuex from 'vuex'

Vue.use(Vuex)

export default new Vuex.Store({
  state: {
    orderType: 'food',
    orders: []
  },
  mutations: {
    CHANGE_ORDER_TYPE(state, type) {
      state.orderType = type
    },

    ADD_ORDER(state, orders) {
      state.orders.push({ ...orders, type: state.orderType })
    }
  },
  getters: {
    filteredOrders({ orders, orderType }) {
      return orders.filter(o => o.type === orderType)
    }
  }
})
