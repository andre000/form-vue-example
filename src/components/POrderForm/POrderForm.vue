<template>
  <validation-observer ref="form-validator" v-slot="{ invalid }">
    <form
      class="form"
      @submit.prevent="handleSubmit"
      @reset.prevent="handleReset"
    >
      <img
        src="@/assets/img/pasteis-img.png"
        class="form__pasteis hide-md hide-sm hide-xs"
      />
      <div class="form__header">
        <span class="form__title">
          Monte aqui o seu cardápio. <br class="hide-xl" />
          <span class="hide-xs">O que está esperando?</span>
        </span>
        <div class="form__switch">
          <p-switch
            :first-option="{ label: 'Comida', value: 'food' }"
            :second-option="{ label: 'Bebida', value: 'drinks' }"
            @input="handleOrderType"
            :value="orderType"
          />
        </div>
      </div>

      <div class="form__body">
        <div class="form__row">
          <validation-provider
            :slim="true"
            :rules="{ required: true, min: 3, max: 60 }"
            v-slot="{ errors }"
            name="'Título do Pedido'"
          >
            <p-input
              class="w-40"
              v-model="name"
              placeholder="Título do Pedido"
              :errorMessage="errors[0]"
            />
          </validation-provider>

          <validation-provider
            :slim="true"
            :rules="{ required: true, min: 3, max: 60 }"
            v-slot="{ errors }"
            name="'Sabor'"
          >
            <p-input
              class="w-40"
              v-model="flavour"
              placeholder="Sabor"
              :errorMessage="errors[0]"
            />
          </validation-provider>

          <validation-provider
            class="w-20"
            :slim="true"
            :rules="{ required: true, min_value: 0.01 }"
            v-slot="{ errors }"
            name="'Preço'"
            ref="price-validator"
          >
            <p-input
              v-model="price"
              @input="handlePriceValidation"
              placeholder="R$"
              is-money
              :errorMessage="!isEmptyForm ? errors[0] : ''"
            />
          </validation-provider>
        </div>

        <div class="form__row">
          <p-text-area
            placeholder="Descrição"
            class="w-100"
            v-model="description"
          />
        </div>

        <div class="form__row">
          <p-image-upload :image="image" @upload="file => (image = file)" />
        </div>
      </div>

      <div class="form__footer">
        <div class="form__actions">
          <p-button color="primary" type="reset">Limpar</p-button>
          <p-button color="secondary" type="submit" :disabled="invalid"
            >Cadastrar</p-button
          >
        </div>
      </div>
    </form>
  </validation-observer>
</template>

<script>
import { mapMutations, mapState } from 'vuex'

import { PInput } from '@/components/PInput'
import { PTextArea } from '@/components/PTextArea'
import { PImageUpload } from '@/components/PImageUpload'
import { PButton } from '@/components/PButton'
import { PSwitch } from '@/components/PSwitch'
import { ValidationProvider, ValidationObserver } from '@/plugins/vee-validate'

export default {
  name: 'PForm',
  components: {
    ValidationProvider,
    ValidationObserver,
    PInput,
    PTextArea,
    PImageUpload,
    PButton,
    PSwitch
  },

  data: () => ({
    name: '',
    flavour: '',
    price: 0,
    description: '',
    image: null
  }),

  computed: {
    isEmptyForm() {
      return Object.keys(this.$data).every(data => !this[data])
    },
    ...mapState(['orderType'])
  },

  methods: {
    handleSubmit() {
      this.ADD_ORDER({ ...this.$data })
      this.handleReset()
    },
    handleOrderType(event) {
      this.CHANGE_ORDER_TYPE(event)
    },
    handleReset() {
      this.name = ''
      this.flavour = ''
      this.price = 0
      this.description = ''
      this.image = null

      this.$refs['form-validator'].reset()
    },

    handlePriceValidation() {
      this.$refs['price-validator']?.validate()
    },

    ...mapMutations(['ADD_ORDER', 'CHANGE_ORDER_TYPE'])
  }
}
</script>

<style lang="scss">
@import '@/assets/scss/variables.scss';

.form {
  width: 60vw;
  box-shadow: 0px 0px 30px $shadow-color;
  border-radius: 20px;
  position: relative;
  background: #ffffff 0% 0% no-repeat padding-box;

  @media #{$screen-xs} {
    width: 80vw;
  }

  .form__pasteis {
    width: 393px;
    height: 316px;
    position: absolute;
    top: calc(-316px / 2);
    clip-path: polygon(0 0, 100% 0, 100% 50%, 0 50%);
    right: 0;
  }

  .form__header {
    background: $secondary;
    padding: 25px 60px 40px 60px;
    position: relative;
    border-radius: 20px 20px 0px 0px;

    @media screen and (max-width: 768px) {
      text-align: center;
      padding: 25px 30px 40px 30px;
    }

    .form__title {
      color: $primary-text;
      font: Bold Italic 24px/29px Roboto;
    }

    .form__switch {
      color: $primary-text;
      position: absolute;
      right: 60px;
      top: 25px;

      @media screen and (max-width: 768px) {
        position: relative;
        right: 0px;
        top: 0px;
        margin-top: 12px;
      }
    }
  }

  .form__body {
    display: flex;
    flex-direction: column;
    padding: 0px 20px 50px 20px;
    margin-top: -20px;
    position: relative;

    .form__row {
      display: flex;
      margin-bottom: 20px;
      position: relative;

      @media #{$screen-xs} {
        &:first-child {
          flex-direction: column;
          .field {
            width: 100%;
            margin: 10px 0px !important;
          }
        }
      }

      .field:not(:first-child):not(:last-child):not(:only-child) {
        margin: 0px 10px;
      }
      .field:first-child {
        margin: 0px 10px 0px 0px;
      }
      .field:last-child {
        margin: 0px 0px 0px 10px;
      }
      .field:only-child {
        margin: 0px;
      }

      textarea {
        height: 80px;
      }

      .image-upload {
        width: 100%;
        height: calc(110px - 40px - 2px);
      }
    }
  }

  .form__footer {
    position: absolute;
    width: 100%;
    bottom: -30px;
    text-align: center;
    @media #{$screen-xs} {
      bottom: -100px;
      display: flex;
      flex-direction: column;
      justify-content: center;
    }

    .button {
      &:first-child {
        margin-right: 35px;
      }
      @media #{$screen-xs} {
        margin: 5px 0px !important;
      }
    }
  }
}
</style>
