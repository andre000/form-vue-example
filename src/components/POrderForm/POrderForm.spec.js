import { mount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import Vuex from 'vuex'

import POrderForm from './POrderForm'
import store from '@/store'

const localVue = createLocalVue()
localVue.use(Vuex)

const mockedOrder = {
  name: 'Suco de Goiaba',
  flavour: 'Goiaba',
  price: 5,
  description: 'Um mero suco de goiaba. Tem sachet de açúcar no balcão...',
  image: { preview: 'image' }
}

jest.useFakeTimers()
async function flush() {
  await flushPromises()
  jest.runAllTimers()
}

describe('POrderForm component', () => {
  let wrapper

  beforeAll(() => {
    wrapper = mount(POrderForm, {
      localVue,
      store,
      sync: false
    })
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a form element', () => {
    expect(wrapper.find('.form').exists()).toBe(true)
  })

  it('should have a disabled submit button', async () => {
    await flush()
    expect(wrapper.find('button[type=submit]:disabled').exists()).toBe(true)
  })

  it('should submit successfully when filled', () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder })
    })

    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      false
    )
  })

  it('should show an error when name field is empty', async () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder, name: '' })
    })

    await wrapperWithValue.vm.$refs['form-validator'].validate()
    await flush()

    expect(
      wrapperWithValue.find('input ~ .field__message--error').text()
    ).toMatch(`O campo 'Título do Pedido' é obrigatório`)
    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      true
    )
  })

  it('should show an error when name field has less than 3 characters', async () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder, name: 'f' })
    })

    await wrapperWithValue.vm.$refs['form-validator'].validate()
    await flush()

    expect(
      wrapperWithValue.find('input ~ .field__message--error').text()
    ).toMatch(`O campo 'Título do Pedido' deve conter pelo menos 3 caracteres`)
    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      true
    )
  })

  it('should show an error when flavour field is empty', async () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder, flavour: '' })
    })

    await wrapperWithValue.vm.$refs['form-validator'].validate()
    await flush()

    expect(
      wrapperWithValue.find('input ~ .field__message--error').text()
    ).toMatch(`O campo 'Sabor' é obrigatório`)
    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      true
    )
  })

  it('should show an error when flavour field has less than 3 characters', async () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder, flavour: 'f' })
    })

    await wrapperWithValue.vm.$refs['form-validator'].validate()
    await flush()

    expect(
      wrapperWithValue.find('input ~ .field__message--error').text()
    ).toMatch(`O campo 'Sabor' deve conter pelo menos 3 caracteres`)
    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      true
    )
  })

  it('should show an error when price field is equal to zero', async () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder, price: 0 })
    })

    await wrapperWithValue.vm.$refs['form-validator'].validate()
    await flush()

    expect(
      wrapperWithValue.find('input ~ .field__message--error').text()
    ).toMatch(`O campo 'Preço' precisa ser 0.01 ou maior`)
    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      true
    )
  })

  it('should reset all values when form is reseted', async () => {
    const wrapperWithValue = mount(POrderForm, {
      localVue,
      store,
      sync: false,
      data: () => ({ ...mockedOrder })
    })

    await wrapperWithValue.find('form').element.reset()

    await wrapperWithValue.vm.$refs['form-validator'].validate()
    await flush()

    expect(wrapperWithValue.find('button[type=submit]:disabled').exists()).toBe(
      true
    )
    expect(wrapperWithValue.vm.$data).toEqual(wrapper.vm.$data)
  })
})
