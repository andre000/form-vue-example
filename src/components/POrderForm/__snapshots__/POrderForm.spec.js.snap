// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POrderForm component should match the snapshot 1`] = `
<span><form class="form"><div class="form__switch"><label for="toggle-button"><span>Comida</span> <input type="checkbox" id="toggle-button" style="display: none;"> <i> ⬅ </i>
<!----> <span>Bebida</span></label></div>
<div class="form__line"><span><div class="field"><input placeholder="Título do Pedido"> <!----></div></span> <span><div class="field"><input placeholder="Sabor"> <!----></div></span> <span><div class="field"><input type="tel" class="v-money"> <!----></div></span></div>
<div class="form__line">
  <div class="text"><textarea></textarea>
    <!---->
  </div>
</div>
<div class="form__line">
  <div class="image-upload"><input type="file" accept=".png, .jpg, .jpeg" name="image-order" class="image-upload__input">
    <!---->
  </div>
</div>
<div class="form__actions"><button class="button button--default" type="reset">Limpar</button> <button class="button button--default" type="submit">Cadastrar</button></div>
</form>
</span>
`;
